### ISSUE TITLE
[Issue]: HIP Graph: Non-VM memory pool path breaks tensor address stability across captures

### ISSUE THREAD

**Issue by sanbuphy** [ROCm-Developer-Tools/clr#280]
## Summary

When `HIP_MEM_POOL_USE_VM` is disabled (which happens at runtime when `HSA_AMD_SYSTEM_INFO_VIRTUAL_MEM_API_SUPPORTED` returns false), `capturehipMallocAsync` uses the `Execute()` path instead of `ReserveAddress()`. This path does not guarantee that tensor addresses remain stable across graph capture/instantiation/replay cycles. PyTorch's `cudagraph_trees` feature depends on this stability and breaks on affected GPUs.

**Affected GPU**: AMD Instinct MI300X (gfx942)  
**ROCm version**: 7.0.51831  
**PyTorch version**: 2.10.0+rocm7.0  

---

### The code path (3 files, 3 lines)

**1. Runtime flag gets disabled** — `rocclr/device/rocm/rocdevice.cpp:1695`

```cpp
info_.virtualMemoryManagement_ = false;
if (HIP_VMEM_MANAGE_SUPPORT) {
    hsa_system_get_info(HSA_AMD_SYSTEM_INFO_VIRTUAL_MEM_API_SUPPORTED,
                        &info_.virtualMemoryManagement_);
}
HIP_MEM_POOL_USE_VM &= info_.virtualMemoryManagement_;  // ← becomes false
```

On MI300X + ROCm 7.0, `HSA_AMD_SYSTEM_INFO_VIRTUAL_MEM_API_SUPPORTED` returns `false`. The flag starts as `true` (from `flags.hpp:203`) but is AND-ed to `false`.

**2. Non-VM allocation path** — `hipamd/src/hip_graph.cpp:999`

```cpp
*dev_ptr = (HIP_MEM_POOL_USE_VM)
    ? mem_alloc_node->ReserveAddress()   // VM: stable virtual address ✅
    : mem_alloc_node->Execute(s);        // NON-VM: eager allocation ❌
```

When VM is off, `Execute(s)` immediately allocates from the memory pool and returns a concrete pointer. During the *next* graph capture, the pool's internal state may have changed (due to previous frees, other allocations, or pool trimming), causing a different address to be returned. The graph was recorded with the original address, so replay reads/writes stale memory.

**3. Free path has the same issue** — `hipamd/src/hip_graph.cpp:1015`

```cpp
if (!HIP_MEM_POOL_USE_VM) {
    mem_free_node->Execute(s);  // eager free, may shift subsequent allocations
}
```

### The symptom in PyTorch

PyTorch's `cudagraph_trees` (`torch/_inductor/cudagraph_trees.py`) creates a private memory pool and records multiple CUDA graphs that share it. It asserts that `static_input` tensors have the same `data_ptr` across captures:

```python
# torch/_inductor/compile_fx.py:1934
assert dst.data_ptr() == src.data_ptr()  # ← FAILS on ROCm
```

This assertion catches the address drift. The old cudagraph implementation (`cudagraphify_impl`) does not assert this and appears to work, though it may have subtle correctness issues from the same root cause.

---

### Minimal reproduction (pure HIP, no PyTorch needed)

```python
import torch

class Net(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = torch.nn.Linear(256, 256)
    def forward(self, x):
        return self.fc(x)

# Compare two torch.compile paths:
# ✅ Works:  mode='default'  (no cudagraphs)
# ❌ Fails:  mode='reduce-overhead' (cudagraph_trees enabled)

model = Net().cuda()
model = torch.compile(model, mode='reduce-overhead')  # ← triggers the bug

# Training loss will stall because graph replay uses stale memory addresses
```

### Isolation evidence

| Test | Result |
|------|:------:|
| `torch.cuda.CUDAGraph` directly | ✅ Works |
| inductor old cudagraph (`triton.cudagraph_trees=False`) | ✅ Works |
| inductor new cudagraph (`triton.cudagraph_trees=True`) | ❌ Breaks |
| `mode='reduce-overhead'` | ❌ Breaks |
| Setting `HIP_MEM_POOL_USE_VM=1` env var | ❌ No effect (overridden by HSA) |

### Loss trace (30-step training)

```
Step    No-compile    reduce-overhead
  1     8.934          8.914          ← graph being captured, OK
  3     8.089          8.391          ← replay begins, divergence starts
 16     7.221          8.421          ← loss completely stalled for 13 steps
 17     7.200          7.965          ← graph re-recorded (guard failure), recovers
 30     6.840          7.214          ← permanently lagging
```

---

### Option A: Make non-VM path address-stable (recommended for `ROCm/clr`)

In `hipamd/src/hip_graph.cpp`, make the non-VM `Execute()` path use a deterministic allocation strategy from the memory pool during graph capture. For example:

- Reserve a base address for the pool at creation time
- Use a bump allocator within the pool for graph capture allocations
- Never reuse freed addresses within the same pool until the graph is destroyed

This ensures that across captures, the same sequence of allocations returns the same sequence of addresses.

### Option B: Enable VM on MI300X (if hardware supports it)

If MI300X (gfx942) hardware actually supports virtual memory management, fix `HSA_AMD_SYSTEM_INFO_VIRTUAL_MEM_API_SUPPORTED` to return `true`. The current ROCm 7.0 HSA driver returns `false`.

### Option C: Document the limitation

If neither A nor B is feasible short-term, document that `HIP_MEM_POOL_USE_VM` may be `false` on some GPUs and that graph memory pool users (like PyTorch) should not depend on address stability without checking this flag.

---

## Impact

- **PyTorch `torch.compile(mode='reduce-overhead')`**: completely broken on affected GPUs
- **Any framework using HIP graphs with shared memory pools**: potentially affected
- **Training convergence**: silently degraded (~2x worse), no crash or error

---

## Environment

```
GPU: AMD Instinct MI300X (gfx942), 205 GB VRAM
ROCm: 7.0.51831
HSA Runtime: included with ROCm 7.0
PyTorch: 2.10.0+rocm7.0
OS: Ubuntu 22.04, Linux 6.1.0-39-amd64
```

---

## Related Repos & Files

| Repo | File | Line | Issue |
|------|------|------|-------|
| `ROCm/clr` | `rocclr/device/rocm/rocdevice.cpp` | 1695 | `HIP_MEM_POOL_USE_VM` disabled at runtime |
| `ROCm/clr` | `hipamd/src/hip_graph.cpp` | 999 | Non-VM `Execute()` path |
| `ROCm/clr` | `hipamd/src/hip_graph.cpp` | 1015 | Non-VM `Execute()` free path |
| `ROCm/clr` | `hipamd/src/hip_mempool_impl.cpp` | — | Pool allocator, no address stability guarantee |
| `ROCm/pytorch` | `torch/_inductor/cudagraph_trees.py` | — | 0 HIP references, assumes CUDA address stability |
| `ROCm/pytorch` | `torch/_inductor/compile_fx.py` | 1934 | `data_ptr` assertion fails |

---

## PyTorch-side workaround (separate PR)

While this HIP issue is being fixed, PyTorch can work around it by falling back to the old cudagraph implementation on ROCm:

```python
# torch/_inductor/compile_fx.py
if config.triton.cudagraph_trees:
    if torch.version.hip:
        cudagraphify_fn = cudagraphify_impl  # old impl, doesn't assert address stability
    else:
        cudagraphify_fn = functools.partial(new_cudagraphify_impl, ...)
```

This workaround will be submitted as a separate PR to `ROCm/pytorch`.

---

*Report prepared: 2026-06-19 | Full investigation log: https://github.com/jingyaogong/minimind/blob/main/FINAL_REPORT.md*

### Operating System

Ubuntu 22.04.5 LTS

### CPU

Intel Xeon Platinum 8568Y+ (160 cores, 2 sockets)

### GPU

AMD Instinct MI300X (gfx942)

### ROCm Version

ROCm 7.0.51831

### (Optional for Linux users) Output of /opt/rocm/bin/rocminfo --support

<details>
<summary>rocminfo --support output</summary>

```
Paste output here
```

</details>

**Comment by sanbuphy** [comment]
Releate Question:https://github.com/pytorch/pytorch/issues/187715
